dali/kernels/signal/window/extract_windows_gpu_test.cu

```diff
@@ -0,0 +1,198 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include <vector>
+#include "dali/core/util.h"
+#include "dali/test/test_tensors.h"
+#include "dali/kernels/signal/window/extract_windows_gpu.cuh"
+#include "dali/kernels/scratch.h"
+#include "dali/kernels/signal/window/window_functions.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+TEST(ExtractWindowsGPU, NonBatchedKernel) {
+  float *in_gpu, *out_gpu;
+  int winlen = 60;
+  int windows = 80;
+  int stride = windows;
+  int step = 10;
+  int length = windows * step - 100;;
+  int center = 5;
+  bool reflect = false;
+  cudaMalloc(&in_gpu, sizeof(float)*length);
+  cudaMalloc(&out_gpu, sizeof(float)*windows*winlen);
+  std::vector<float> in(length), out(windows*winlen);
+
+  for (int i = 0; i < length; i++) {
+    in[i] = i + 1000;
+  }
+
+  cudaMemcpy(in_gpu, in.data(), sizeof(float)*length, cudaMemcpyHostToDevice);
+  cudaMemset(out_gpu, 0, sizeof(float)*windows*winlen);
+  int xblocks = div_ceil(length, 32);
+  int yblocks = div_ceil(winlen, 32);
+  window::ExtractWindowsKernel<<<dim3(xblocks, yblocks), dim3(32, 32)>>>(
+    out_gpu, windows, stride, in_gpu, length, nullptr, winlen, center, step, reflect);
+  cudaMemcpy(out.data(), out_gpu, sizeof(float)*winlen*windows, cudaMemcpyDeviceToHost);
+  cudaDeviceSynchronize();
+
+  for (int w = 0; w < windows; w++) {
+    for (int i = 0; i < winlen; i++) {
+      int idx = w * step + i - center;
+      if (reflect) {
+        for (;;) {
+          if (idx < 0)
+            idx = -idx;
+          else if (idx >= length)
+            idx = 2*(length-1) - idx;
+          else
+            break;
+        }
+      }
+      float ref = idx >= 0 && idx < length ? in[idx] : 0;
+      EXPECT_EQ(out[w + i*stride], ref)
+        << "@ window = " << w << ", index = " << i;
+    }
+  }
+
+  if (HasFailure()) {
+    std::cout << "Debug: Extract window actual output:\n";
+    for (int i = 0; i < winlen; i++) {
+      for (int j = 0; j < windows; j++) {
+        std::cout << out[i*stride+j] << " ";
+      }
+      std::cout << "\n";
+    }
+    std::cout << std::flush;
+  }
+
+  cudaFree(in_gpu);
+  cudaFree(out_gpu);
+}
+
+
+
+void TestBatchedExtract(bool concatenate, Padding padding, span<const float> window) {
+  ExtractWindowsGPUImpl<float, float> extract;
+
+  TensorListShape<1> lengths({ TensorShape<1>{5}, TensorShape<1>{305}, TensorShape<1>{157} });
+  int N = lengths.num_samples();
+
+  ptrdiff_t total_length = 0;
+  for (int i = 0; i < N; i++) {
+    total_length += lengths[i][0];
+  }
+
+  TestTensorList<float, 1> in_list;
+  in_list.reshape(lengths);
+  auto in_cpu = in_list.cpu();
+  for (int i = 0; i < N; i++) {
+    for (int j = 0; j < lengths[i][0]; j++)
+      in_cpu[i].data[j] = 1000*(i+1)+j;
+  }
+
+  ExtractWindowsArgs args;
+  args.window_length = window.empty() ? 55 : window.size();
```

Comment:
suggestion: Those tests could be parametrized. At the very least, those magic numbers could be declared constants at the top of the test